### PR TITLE
avm2: Reimplement Matrix3D.recompose() as a native method

### PR DESCRIPTION
**Pull request by kjarosh**
## Description

This patch removes the AS-based Matrix3D.recompose() implementation and
makes it a native method instead. The implementation is based on the AS
one, but edge cases are handled properly.

Based on those edge cases, we should be multiplying matrices instead of
doing low-level operations on raw data (observable via e.g. NaN
propagation).

## Testing

Existing tests + test included.

## Checklist

- [x] I, a human, have self-reviewed this PR and fully understand the changes within.
- [x] I have [made or updated tests](https://github.com/ruffle-rs/ruffle/blob/master/CONTRIBUTING.md#test-guidelines) where possible.
- [x] All of my commits are properly scoped, compile successfully, and pass all tests.
- [x] This PR does not make sense to split up into smaller PRs.
- [x] An LLM was involved in the authoring of this code.

### CODE REVIEW

**Review by SuchAFuriousDeath — APPROVED** [review]

File: core/src/avm2/globals/flash/geom/matrix_3d.rs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    #[avm2_variant("eulerAngles")]
+    EulerAngles,
+    #[avm2_variant("axisAngle")]
+    AxisAngle,
+    #[avm2_variant("quaternion")]
+    Quaternion,
+}
+
+/// Reads the `x`, `y`, `z` and `w` components of a `Vector3D`, or `None` if it's null.
+fn read_vector3d(value: Value<'_>) -> Option<[f64; 4]> {
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/geom/matrix_3d.rs]: IMO it's nicer for this method to take `Option<Object>`- then its caller doesn't need to manifest a temporary `Value::Undefined`.

Comment (kjarosh) [reply core/src/avm2/globals/flash/geom/matrix_3d.rs]: Done
